### PR TITLE
Binder (Jupyter Notebook) Example w/ GLVis

### PR DESCRIPTION
**Pull request by tomstitt**
This adds a _binder_ directory with configuration for https://mybinder.org and a _notebooks_ directory with a C++ notebook (xeus-cling) example modeled off of ex1 that includes an interactive glvis widget via https://github.com/GLVis/xeus-glvis.

The xeus-glvis interface is still in flux so if you have recommendations please let me know.

I would like to get rid of the `#pragma cling` statements that include the binder-specific paths but I'm not sure how to do that yet (they shouldn't hurt a local run though).

The way the binder environment is setup is that both mfem and xeus-glvis will build and/or installed from their main branches whenever anything is merged to `mfem@master`; this means that the container build is a bit slow for the first person to open a binder link after a push to `master` since we have to build `libmfem`. An alternative would be to just use the numbered mfem releases but that would mean new features won't be available.

To try the example: https://mybinder.org/v2/gh/mfem/mfem/binder?filepath=examples%2Fjupyter%2Fex.ipynb

We can add a badge if we want:

[Binder](https://mybinder.org/v2/gh/mfem/mfem/binder?filepath=examples%2Fjupyter%2Fex.ipynb)

- [ ] Update link in readme badge to point at `master` right before merging

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#2477](https://github.com/mfem/mfem/pull/2477) | @tomstitt | @v-dobrev | @tzanio + @rw-anderson + @artv3 | 08/19/21 | 09/15/21 | 09/26/21 | |

### CODE REVIEW

**Comment by tomstitt** [comment]
One minor issue is that while `display` is the method in _pyglvis_ to show a new figure it is not usable in the c++ because it is reserved by one of _xeus-glvis_'s parent classes. If we want consistency we can change both to something like `draw` or `show`, thoughts?

**Comment by tzanio** [comment]
Either one is fine with me. Another option maybe is `plot`.

**Comment by v-dobrev** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Comment by v-dobrev** [comment]
Maybe the `notebooks` directory can be moved under `examples`.

Also, can the `binder` directory be moved under `config`?

**Comment by tzanio** [comment]
This is really awesome, thanks Tom!

**Comment by tzanio** [comment]
6️⃣ Regarding `xeus-glvis` -- I wonder if it has to be in a separate repo, or can we just add `xglvis.hpp` to the main GLVis repo (I'm happy either way)

**Comment by camierjs** [comment]
This is really neat! Works great!

**Comment by rw-anderson** [comment]
This is amazing. Now I can prototype my code in a python notebook first, then a C++ notebook, then move it into production. :)

File: notebooks/ex.ipynb [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+   "source": [
+    "#pragma cling load(\"libmfem\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "explicit-vacuum",
+   "metadata": {},
+   "source": [
+    "## Now an example"
```
Comment (rw-anderson) [inline notebooks/ex.ipynb:49]: Maybe show off the rich documentation of the notebook a little more with something like this:

This is the simplest MFEM example and a good starting point for new users. The example demonstrates the use of MFEM to define and solve an $H^1$ finite element discretization of the Laplace problem 
$$
\nabla^2 u = 1
$$
with homogeneous Dirichlet boundary conditions $u=0$.

The example illustrates the use of the basic MFEM classes for defining the mesh, finite element space, as well as linear and bilinear forms corresponding to the left-hand side and right-hand side of the discrete linear system.

**Review by rw-anderson — APPROVED** [review]
Looks great.

I suggested adding some equations, just to show off the rich documentation capability of notebooks as an advantage over other way to write code.

File: notebooks/ex.ipynb (hunk 1 above)
Comment (tzanio) [reply notebooks/ex.ipynb:49]: I agree. Feel free to just add it in and commit @rw-anderson.

**Comment by tomstitt** [comment]
> @v-dobrev wrote:
> Maybe the `notebooks` directory can be moved under `examples`.
> 
> Also, can the `binder` directory be moved under `config`?

> @tzanio wrote:
> 1️⃣ Can we put all of this under examples/jupyter. I think it will be cleaner if we can.

We can definitely move the examples to _example/jupyter_ or _examples/notebooks_, which one do we want?

The configuration files needs to be in either the root, in _binder_, or in _.binder_ (https://repo2docker.readthedocs.io/en/latest/specification.html), do y'all prefer one of the others?

**Comment by tzanio** [comment]
> We can definitely move the examples to example/jupyter or examples/notebooks, which one do we want?
>
> The configuration files needs to be in either the root, in binder, or in .binder (https://repo2docker.readthedocs.io/en/latest/specification.html), do y'all prefer one of the others?

I prefer `examples/jupyter` and `.binder`

**Comment by tomstitt** [comment]
> 3️⃣ I agree that the /home/jovyan/local/ lines at the beginning look strange. It will be great if we can remove them.

I was able to install things in such a way that the `#include`s work without a cling pragma but the lib isn't as nice. Something in the cling runtime updates the library path after a `#include` of a tpl such that we _could_ do this:

```
#Include <glvis/xglvis.hpp>
#pragma cling load("libmfem")
#include <mfem.hpp>
```

but that looks weird so I just changed the top `load` to the full path of `libmfem`. I will see if we can do better and get to just `#pragma cling load("mfem")`

**Comment by tomstitt** [comment]
> 2️⃣ What will it take to run this locally? Should we include a README.md with details how to do that? We can put the binder button there

I made an initial readme in _examples/jupyter_ and included the binder link and local install/run instructions

**Comment by tomstitt** [comment]
> 5️⃣ Are there any restriction to what code we can put in the Notebook? Can we mix this with pymfem code?

MPI code will most likely not work because each rank will try to launch its own kernel; instead we'd need them all to follow the lead of a single rank. We can probably write our own one with mpi support and I'm planning on a mpi-xeus implementation (xeus is the jupyter protocol implementation that xeus-cling is built on) for a different project so we can try to expend that if there is an mpi use-case

We cannot mix PyMFEM code at the moment, that would require an embedded python runtime along with some interoperability layer but it's definitely possible 

> 6️⃣ Regarding xeus-glvis -- I wonder if it has to be in a separate repo, or can we just add xglvis.hpp to the main GLVis repo (I'm happy either way)

Moving it to the main GLVis repo is fine with me as long as we're okay having an `install` target that doesn't actually build `libglvis`. I was thinking we would need to add some stuff to `xeus-glvis` to handle the widget frontend but I think just depending on a `pyglvis` is actually a okay approach, so the repo will remain pretty minimal (an argument for moving?)

**Comment by tzanio** [comment]
Thanks for the changes and the explanation @tomstitt!

Regarding `xeus-glvis` -- it is totally up to you. From what I saw, I thought it was just a header, but maybe I misunderstood. Can you clarify why we need the `install` target you mentioned above?

**Comment by tzanio** [comment]
Can you please mention this in `CHANGELOG`?

**Comment by tzanio** [comment]
Update `CHANGELOG` and I'll approve :)

**Comment by tomstitt** [comment]
> Update CHANGELOG and I'll approve 😁

done!

**Review by tzanio — COMMENTED** [review]
Amazing stuff, thank you @tomstitt !  😲 🚀 👍

**Review by tzanio — APPROVED** [review]

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Pull request merged**